app/services/service_listeners/attachment_replacement_id_updater.rb

```diff
@@ -0,0 +1,38 @@
+module ServiceListeners
+  class AttachmentReplacementIdUpdater
+    include Rails.application.routes.url_helpers
+    include PublicDocumentRoutesHelper
+
+    attr_reader :attachment_data, :queue
+
+    def initialize(attachment_data, queue: nil)
+      @attachment_data = attachment_data
+      @queue = queue
+    end
+
+    def update!
+      return unless attachment_data.present?
+
+      return unless attachment_data.replaced_by.present?
+      replacement = attachment_data.replaced_by
+
+      enqueue_job(attachment_data.file, replacement.file)
+      if attachment_data.pdf?
+        enqueue_job(attachment_data.file.thumbnail, replacement.file.thumbnail)
```

Comment:
Is it possible to replace a PDF with something other than a PDF? And if so, will this do something sensible?

Reply:
@chrisroos:

It does seem to be possible to to replace a PDF with a non-PDF. I've just tried this in integration and the original PDF thumbnail gets redirected to the replacement attachment URL (not its thumbnail, which doesn't exist anyway).

I think this is because once the PDF attachment has been replaced, and a request is made for the thumbnail, `AttachmentsController#show` will lookup the `AttachmentData` for the original PDF (which is the same for its thumbnail) and within `AttachmentsController#fail` it will redirect to the replacement attachment URL (not it's non-existent thumbnail). I suspect this scenario was not contemplated when this functionality was added, because the behaviour doesn't seem very sensible!

I'm now going to investigate what happens in this scenario with my changes in place.

Note that in doing the testing above in integration, I can see that the host is being set correctly (to `www-origin`) for the redirect - I had previously seen this being set to `whitehall-admin` vs `www` on my dev VM. So I assume that is only a problem on the dev VM.

Reply:
@chrisroos I've addressed this in a new commit: 385d7bdebe4c8558569b813281daa185525799c4.